File: packages/yuukei-intelligence/src/output.mjs
export function silentOutput() {
  return { speak: false };
}

export function unknownChoiceOutput() {
  return { choice: "不明" };
}

export function unknownExtractOutput() {
  return { found: false, value: "不明" };
}

export function memoryIndexFailureOutput() {
  return { indexed: false };
}

export function moodEvaluateFailureOutput() {
  return { mood: "ふつう", talkDesire: 50, topic: "" };
}

export function normalizeOutput(value, maxLength = 120) {
  if (!value || typeof value !== "object" || typeof value.speak !== "boolean") {
    return silentOutput();
  }
  if (!value.speak) {
    return silentOutput();
  }
  const text = typeof value.text === "string" ? value.text.trim() : "";
  if (!text) {
    return silentOutput();
  }
  const output = {
    speak: true,
    text: text.slice(0, Math.max(1, maxLength))
  };
  if (typeof value.expression === "string" && value.expression.trim()) {
    output.expression = value.expression.trim();
  }
  if (typeof value.motion === "string" && value.motion.trim()) {
    output.motion = value.motion.trim();
  }
  return output;
}

export function normalizeInterpretOutput(value, choices = []) {
  if (!value || typeof value !== "object" || typeof value.choice !== "string") {
    return unknownChoiceOutput();
  }
  const choice = value.choice.trim();
  const allowed = new Set([...choices, "不明"]);
  if (!allowed.has(choice)) {
    return unknownChoiceOutput();
  }
  const output = { choice };
  if (typeof value.confidence === "number" && Number.isFinite(value.confidence)) {
    output.confidence = value.confidence;
  }
  return output;
}

export function normalizeExtractOutput(value) {
  if (!value || typeof value !== "object" || value.found !== true) {
    return unknownExtractOutput();
  }
  const text = typeof value.value === "string" ? value.value.trim() : "";
  if (!text || text.length > 100) {
    return unknownExtractOutput();
  }
  return { found: true, value: text };
}

export function parseJsonOutput(text, maxLength) {
  if (typeof text !== "string" || !text.trim()) {
    return silentOutput();
  }
  const candidates = [stripCodeFence(text), extractJsonObject(text)].filter(Boolean);
  const uniqueCandidates = [...new Set(candidates)];
  let lastError;
  for (const candidate of uniqueCandidates) {
    try {
      return normalizeOutput(JSON.parse(candidate), maxLength);
    } catch (error) {
      lastError = error;
    }
  }
  console.error(
    `yuukei-intelligence: failed to parse provider JSON: ${lastError?.message ?? "no JSON object found"}`
  );
  return silentOutput();
}

export function parseJsonInterpretOutput(text, choices) {
  if (typeof text !== "string" || !text.trim()) {
    return unknownChoiceOutput();
  }
  const candidates = [stripCodeFence(text), extractJsonObject(text)].filter(Boolean);
  const uniqueCandidates = [...new Set(candidates)];
  let lastError;
  for (const candidate of uniqueCandidates) {
    try {
      return normalizeInterpretOutput(JSON.parse(candidate), choices);
    } catch (error) {
      lastError = error;
    }
  }
  console.error(
    `yuukei-intelligence: failed to parse provider JSON: ${lastError?.message ?? "no JSON object found"}`
  );
  return unknownChoiceOutput();
}

export function parseJsonExtractOutput(text) {
  if (typeof text !== "string" || !text.trim()) {
    return unknownExtractOutput();
  }
  const candidates = [stripCodeFence(text), extractJsonObject(text)].filter(Boolean);
  const uniqueCandidates = [...new Set(candidates)];
  let lastError;
  for (const candidate of uniqueCandidates) {
    try {
      return normalizeExtractOutput(JSON.parse(candidate));
    } catch (error) {
      lastError = error;
    }
  }
  console.error(
    `yuukei-intelligence: failed to parse extract JSON: ${lastError?.message ?? "no JSON object found"}`
  );
  return unknownExtractOutput();
}

export function parseJsonMemoryIndexOutput(text) {
  if (typeof text !== "string" || !text.trim()) {
    return null;
  }
  const candidates = [stripCodeFence(text), extractJsonObject(text)].filter(Boolean);
  const uniqueCandidates = [...new Set(candidates)];
  let lastError;
  for (const candidate of uniqueCandidates) {
    try {
      return normalizeMemoryIndexSummary(JSON.parse(candidate));
    } catch (error) {
      lastError = error;
    }
  }
  console.error(
    `yuukei-intelligence: failed to parse memory index JSON: ${lastError?.message ?? "no JSON object found"}`
  );
  return null;
}

export function parseJsonMoodEvaluateOutput(text) {
  if (typeof text !== "string" || !text.trim()) {
    return moodEvaluateFailureOutput();
  }
  const candidates = [stripCodeFence(text), extractJsonObject(text)].filter(Boolean);
  const uniqueCandidates = [...new Set(candidates)];
  let lastError;
  for (const candidate of uniqueCandidates) {
    try {
      return normalizeMoodEvaluateOutput(JSON.parse(candidate));
    } catch (error) {
      lastError = error;
    }
  }
  console.error(
    `yuukei-intelligence: failed to parse mood JSON: ${lastError?.message ?? "no JSON object found"}`
  );
  return moodEvaluateFailureOutput();
}

export function normalizeMoodEvaluateOutput(value) {
  if (!value || typeof value !== "object") {
    return moodEvaluateFailureOutput();
  }
  const mood = normalizeMoodWord(value.mood);
  const talkDesire = clampTalkDesire(value.talkDesire);
  const topic = typeof value.topic === "string" ? value.topic.trim().slice(0, 80) : "";
  return { mood, talkDesire, topic };
}

export function normalizeMoodWord(value) {
  const word = typeof value === "string" ? value.trim() : "";
  return ["ふつう", "うれしい", "たいくつ", "さみしい", "心配", "ねむい"].includes(word)
    ? word
    : "ふつう";
}

function clampTalkDesire(value) {
  const number = Number(value);
  if (!Number.isFinite(number)) {
    return 50;
  }
  return Math.min(100, Math.max(0, Math.trunc(number)));
}

export function normalizeMemoryIndexSummary(value) {
  if (!value || typeof value !== "object") {
    return null;
  }
  const diary = typeof value.diary === "string" ? value.diary.trim() : "";
  const newFacts = Array.isArray(value.newFacts)
    ? value.newFacts
        .filter((fact) => typeof fact === "string")
        .map((fact) => fact.trim())
        .filter(Boolean)
        .slice(0, 5)
    : [];
  if (!diary && newFacts.length === 0) {
    return null;
  }
  return { diary, newFacts };
}

function stripCodeFence(text) {
  const trimmed = text.trim();
  const match = trimmed.match(/^```(?:json)?\s*([\s\S]*?)\s*```$/i);
  return match ? match[1].trim() : trimmed;
}

function extractJsonObject(text) {
  const source = stripCodeFence(text);
  const start = source.indexOf("{");
  if (start < 0) {
    return null;
  }
  let depth = 0;
  let inString = false;
  let escaped = false;
  for (let index = start; index < source.length; index += 1) {
    const char = source[index];
    if (inString) {
      if (escaped) {
        escaped = false;
      } else if (char === "\\") {
        escaped = true;
      } else if (char === "\"") {
        inString = false;
      }
      continue;
    }
    if (char === "\"") {
      inString = true;
    } else if (char === "{") {
      depth += 1;
    } else if (char === "}") {
      depth -= 1;
      if (depth === 0) {
        return source.slice(start, index + 1);
      }
    }
  }
  return null;
}

export function capabilityResult(invocation, output, metadata = {}) {
  const capability = invocation?.capability ?? "dialogue.generate";
  let normalizedOutput;
  if (capability === "dialogue.interpret") {
    normalizedOutput = normalizeInterpretOutput(output, invocation?.input?.choices);
  } else if (capability === "dialogue.extract") {
    normalizedOutput = normalizeExtractOutput(output);
  } else if (capability === "memory.index") {
    normalizedOutput = normalizeMemoryIndexCapabilityOutput(output);
  } else if (capability === "memory.list") {
    normalizedOutput = normalizeMemoryListCapabilityOutput(output);
  } else if (capability === "memory.retrieve") {
    normalizedOutput = normalizeMemoryRetrieveCapabilityOutput(output);
  } else if (capability === "memory.update") {
    normalizedOutput = normalizeMemoryUpdateCapabilityOutput(output);
  } else if (capability === "memory.forget") {
    normalizedOutput = normalizeMemoryForgetCapabilityOutput(output);
  } else if (capability === "mood.evaluate") {
    normalizedOutput = normalizeMoodEvaluateOutput(output);
  } else {
    normalizedOutput = normalizeOutput(output, invocation?.input?.constraints?.maxLength);
  }
  return {
    invocationId: invocation?.id ?? "",
    extensionId: "yuukei-intelligence",
    capability,
    output: normalizedOutput,
    metadata
  };
}

function normalizeMemoryIndexCapabilityOutput(value) {
  if (!value || typeof value !== "object") {
    return memoryIndexFailureOutput();
  }
  const output = { indexed: value.indexed === true };
  if (typeof value.noteCount === "number" && Number.isFinite(value.noteCount)) {
    output.noteCount = Math.max(0, Math.trunc(value.noteCount));
  }
  return output;
}

function normalizeMemoryListCapabilityOutput(value) {
  if (!value || typeof value !== "object") {
    return { facts: [], episodes: [], episodeTotal: 0 };
  }
  const facts = Array.isArray(value.facts)
    ? value.facts
        .filter((fact) => fact && typeof fact === "object")
        .map((fact) => ({
          id: typeof fact.id === "string" ? fact.id : "",
          text: typeof fact.text === "string" ? fact.text.trim() : "",
          createdAt: typeof fact.createdAt === "string" ? fact.createdAt : "",
          updatedAt: typeof fact.updatedAt === "string" ? fact.updatedAt : ""
        }))
        .filter((fact) => fact.id && fact.text)
    : [];
  const episodes = Array.isArray(value.episodes)
    ? value.episodes
        .filter((episode) => episode && typeof episode === "object")
        .map((episode) => ({
          id: typeof episode.id === "string" ? episode.id : "",
          text: typeof episode.text === "string" ? episode.text.trim() : "",
          timestamp: typeof episode.timestamp === "string" ? episode.timestamp : ""
        }))
        .filter((episode) => episode.id && episode.text)
    : [];
  return {
    facts,
    episodes,
    episodeTotal:
      typeof value.episodeTotal === "number" && Number.isFinite(value.episodeTotal)
        ? Math.max(0, Math.trunc(value.episodeTotal))
        : episodes.length
  };
}

function normalizeMemoryRetrieveCapabilityOutput(value) {
  if (!value || typeof value !== "object" || !Array.isArray(value.memories)) {
    return { memories: [] };
  }
  return {
    memories: value.memories
      .filter((memory) => memory && typeof memory === "object")
      .map((memory) => {
        const text = typeof memory.text === "string" ? memory.text.trim() : "";
        const kind = memory.kind === "episode" ? "episode" : "fact";
        const output = { text, kind };
        if (kind === "episode" && typeof memory.date === "string" && memory.date.trim()) {
          output.date = memory.date.trim();
        }
        return output;
      })
      .filter((memory) => memory.text)
  };
}

function normalizeMemoryUpdateCapabilityOutput(value) {
  return { updated: Boolean(value && typeof value === "object" && value.updated === true) };
}

function normalizeMemoryForgetCapabilityOutput(value) {
  if (!value || typeof value !== "object") {
    return { removedFacts: 0, removedEpisodes: 0 };
  }
  return {
    removedFacts:
      typeof value.removedFacts === "number" && Number.isFinite(value.removedFacts)
        ? Math.max(0, Math.trunc(value.removedFacts))
        : 0,
    removedEpisodes:
      typeof value.removedEpisodes === "number" && Number.isFinite(value.removedEpisodes)
        ? Math.max(0, Math.trunc(value.removedEpisodes))
        : 0
  };
}

File: packages/yuukei-intelligence/src/output.rs
use serde_json::{json, Map, Value};

pub fn capability_result(invocation: &Value, output: Value, metadata: Value) -> Value {
    let capability = invocation
        .get("capability")
        .and_then(Value::as_str)
        .unwrap_or("dialogue.generate");
    let normalized = match capability {
        "dialogue.interpret" => normalize_interpret(
            &output,
            invocation
                .pointer("/input/choices")
                .and_then(Value::as_array)
                .map(Vec::as_slice)
                .unwrap_or(&[]),
        ),
        "dialogue.extract" => normalize_extract(&output),
        "memory.index" => normalize_memory_index_capability(&output),
        "memory.list" => normalize_memory_list_capability(&output),
        "memory.retrieve" => normalize_memory_retrieve_capability(&output),
        "memory.update" => json!({ "updated": output.get("updated") == Some(&Value::Bool(true)) }),
        "memory.forget" => json!({
            "removedFacts": non_negative_integer(output.get("removedFacts")),
            "removedEpisodes": non_negative_integer(output.get("removedEpisodes"))
        }),
        "mood.evaluate" => normalize_mood(&output),
        _ => normalize_dialogue(
            &output,
            invocation
                .pointer("/input/constraints/maxLength")
                .and_then(Value::as_u64)
                .unwrap_or(120) as usize,
        ),
    };
    json!({
        "invocationId": invocation.get("id").and_then(Value::as_str).unwrap_or(""),
        "extensionId": "yuukei-intelligence",
        "capability": capability,
        "output": normalized,
        "metadata": metadata
    })
}

pub fn silent() -> Value {
    json!({ "speak": false })
}

pub fn unknown_choice() -> Value {
    json!({ "choice": "不明" })
}

pub fn unknown_extract() -> Value {
    json!({ "found": false, "value": "不明" })
}

pub fn mood_failure() -> Value {
    json!({ "mood": "ふつう", "talkDesire": 50, "topic": "" })
}

pub fn parse_dialogue(text: &str, max_length: usize) -> Value {
    parse_object(text)
        .map(|value| normalize_dialogue(&value, max_length))
        .unwrap_or_else(silent)
}

pub fn parse_interpret(text: &str, choices: &[Value]) -> Value {
    parse_object(text)
        .map(|value| normalize_interpret(&value, choices))
        .unwrap_or_else(unknown_choice)
}

pub fn parse_extract(text: &str) -> Value {
    parse_object(text)
        .map(|value| normalize_extract(&value))
        .unwrap_or_else(unknown_extract)
}

pub fn parse_memory_summary(text: &str) -> Option<Value> {
    let value = parse_object(text)?;
    let diary = value
        .get("diary")
        .and_then(Value::as_str)
        .unwrap_or("")
        .trim();
    let new_facts = value
        .get("newFacts")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(Value::as_str)
        .map(str::trim)
        .filter(|fact| !fact.is_empty())
        .take(5)
        .map(|fact| Value::String(fact.to_string()))
        .collect::<Vec<_>>();
    if diary.is_empty() && new_facts.is_empty() {
        return None;
    }
    Some(json!({ "diary": diary, "newFacts": new_facts }))
}

pub fn parse_mood(text: &str) -> Value {
    parse_object(text)
        .map(|value| normalize_mood(&value))
        .unwrap_or_else(mood_failure)
}

fn normalize_dialogue(value: &Value, max_length: usize) -> Value {
    if value.get("speak") != Some(&Value::Bool(true)) {
        return silent();
    }
    let text = value
        .get("text")
        .and_then(Value::as_str)
        .unwrap_or("")
        .trim();
    if text.is_empty() {
        return silent();
    }
    // Expressions and motions are intentionally not accepted from the model.
    // Daihon and the renderer remain authoritative for those asset names.
    json!({
        "speak": true,
        "text": text.chars().take(max_length.max(1)).collect::<String>()
    })
}

fn normalize_interpret(value: &Value, choices: &[Value]) -> Value {
    let choice = value
        .get("choice")
        .and_then(Value::as_str)
        .unwrap_or("")
        .trim();
    let allowed = choice == "不明"
        || choices
            .iter()
            .filter_map(Value::as_str)
            .any(|candidate| candidate == choice);
    if !allowed {
        return unknown_choice();
    }
    let mut output = Map::from_iter([("choice".to_string(), json!(choice))]);
    if let Some(confidence) = value.get("confidence").and_then(Value::as_f64) {
        if confidence.is_finite() {
            output.insert("confidence".to_string(), json!(confidence));
        }
    }
    Value::Object(output)
}

fn normalize_extract(value: &Value) -> Value {
    if value.get("found") != Some(&Value::Bool(true)) {
        return unknown_extract();
    }
    let text = value
        .get("value")
        .and_then(Value::as_str)
        .unwrap_or("")
        .trim();
    if text.is_empty() || text.chars().count() > 100 {
        return unknown_extract();
    }
    json!({ "found": true, "value": text })
}

fn normalize_mood(value: &Value) -> Value {
    let mood = match value.get("mood").and_then(Value::as_str).map(str::trim) {
        Some(value)
            if [
                "ふつう",
                "うれしい",
                "たいくつ",
                "さみしい",
                "心配",
                "ねむい",
            ]
            .contains(&value) =>
        {
            value
        }
        _ => "ふつう",
    };
    let talk_desire = value
        .get("talkDesire")
        .and_then(Value::as_f64)
        .filter(|number| number.is_finite())
        .map(|number| number.trunc().clamp(0.0, 100.0) as i64)
        .unwrap_or(50);
    let topic = value
        .get("topic")
        .and_then(Value::as_str)
        .unwrap_or("")
        .trim()
        .chars()
        .take(80)
        .collect::<String>();
    json!({ "mood": mood, "talkDesire": talk_desire, "topic": topic })
}

fn normalize_memory_index_capability(value: &Value) -> Value {
    let mut output = Map::from_iter([(
        "indexed".to_string(),
        json!(value.get("indexed") == Some(&Value::Bool(true))),
    )]);
    if let Some(note_count) = value.get("noteCount") {
        output.insert(
            "noteCount".to_string(),
            json!(non_negative_integer(Some(note_count))),
        );
    }
    Value::Object(output)
}

fn normalize_memory_list_capability(value: &Value) -> Value {
    let facts = value
        .get("facts")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|fact| {
            let id = fact.get("id")?.as_str()?.to_string();
            let text = fact.get("text")?.as_str()?.trim().to_string();
            (!id.is_empty() && !text.is_empty()).then(|| {
                json!({
                    "id": id,
                    "text": text,
                    "createdAt": fact.get("createdAt").and_then(Value::as_str).unwrap_or(""),
                    "updatedAt": fact.get("updatedAt").and_then(Value::as_str).unwrap_or("")
                })
            })
        })
        .collect::<Vec<_>>();
    let episodes = value
        .get("episodes")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|episode| {
            let id = episode.get("id")?.as_str()?.to_string();
            let text = episode.get("text")?.as_str()?.trim().to_string();
            (!id.is_empty() && !text.is_empty()).then(|| {
                json!({
                    "id": id,
                    "text": text,
                    "timestamp": episode.get("timestamp").and_then(Value::as_str).unwrap_or("")
                })
            })
        })
        .collect::<Vec<_>>();
    let total = value
        .get("episodeTotal")
        .map(|value| non_negative_integer(Some(value)))
        .unwrap_or(episodes.len() as i64);
    json!({ "facts": facts, "episodes": episodes, "episodeTotal": total })
}

fn normalize_memory_retrieve_capability(value: &Value) -> Value {
    let memories = value
        .get("memories")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|memory| {
            let text = memory.get("text")?.as_str()?.trim();
            if text.is_empty() {
                return None;
            }
            let kind = if memory.get("kind").and_then(Value::as_str) == Some("episode") {
                "episode"
            } else {
                "fact"
            };
            let mut result = Map::from_iter([
                ("text".to_string(), json!(text)),
                ("kind".to_string(), json!(kind)),
            ]);
            if kind == "episode" {
                if let Some(date) = memory.get("date").and_then(Value::as_str) {
                    if !date.trim().is_empty() {
                        result.insert("date".to_string(), json!(date.trim()));
                    }
                }
            }
            Some(Value::Object(result))
        })
        .collect::<Vec<_>>();
    json!({ "memories": memories })
}

fn parse_object(text: &str) -> Option<Value> {
    serde_json::from_str::<Value>(text.trim())
        .ok()
        .filter(Value::is_object)
}

fn non_negative_integer(value: Option<&Value>) -> i64 {
    value
        .and_then(Value::as_f64)
        .filter(|number| number.is_finite())
        .map(|number| number.trunc().max(0.0) as i64)
        .unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dialogue_ignores_untrusted_asset_names() {
        let output = parse_dialogue(
            r#"{"speak":true,"text":"こんにちは","expression":"unknown","motion":"unknown"}"#,
            120,
        );
        assert_eq!(output, json!({ "speak": true, "text": "こんにちは" }));
    }

    #[test]
    fn interpret_rejects_choices_outside_daihon() {
        let output = parse_interpret(r#"{"choice":"たぶん"}"#, &[json!("はい"), json!("いいえ")]);
        assert_eq!(output, unknown_choice());
    }

    #[test]
    fn rejects_json_embedded_in_surrounding_text() {
        assert_eq!(
            parse_extract("answer: {\"found\":true,\"value\":\"あんぱん\"} done"),
            unknown_extract()
        );
    }
}
